include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+        {
+            pgtOffset = alloc(_img, _subres, *it);
+            if ((pgtOffset==page_tab_offset_invalid()).all())
+                continue;
+        }
+        if ((pgtOffset==page_tab_offset_invalid()).all())
+        {
+            if (m_freePageTableLayerIDs.empty())
+                return STextureData::invalid();
+            const uint32_t pgtLayer = m_freePageTableLayerIDs.top();
+            m_freePageTableLayerIDs.pop();
+            pgtOffset = alloc(_img, _subres, pgtLayer);
+            if ((pgtOffset==page_tab_offset_invalid()).all())//this would be super weird but let's check
+                return STextureData::invalid();
+            storage->addPageTableLayer(pgtLayer);
+
+            (*m_layerToViewIndexMapping)[pgtLayer] = smplrIndex;
+        }
+
+        const auto extent = _img->getCreationParameters().extent;
+
+        const uint32_t levelsTakingAtLeastOnePageCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+        const uint32_t levelsToPack = std::min(_subres.levelCount, m_pageTable->getCreationParameters().mipLevels+m_pgSzxy_log2);
+
+        uint32_t miptailPgAddr = SPhysPgOffset::invalid_addr;
+
+        using phys_pg_addr_alctr_t = ICPUVTResidentStorage::phys_pg_addr_alctr_t;
+        //TODO up to this line, it's kinda common code for CPU and GPU, refactor later
+
+        //fill page table and pack present mips into physical addr texture
+        CFillImageFilter::state_type fill;
+        fill.outImage = m_pageTable.get();
+        fill.outRange.extent = { 1u,1u,1u };
+        fill.subresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        fill.subresource.baseArrayLayer = pgtOffset.z;
+        fill.subresource.layerCount = 1u;
+        for (uint32_t i = 0u; i < levelsToPack; ++i)
+        {
+            const uint32_t w = neededPageCountForSide(extent.width, _subres.baseMipLevel+i);
+            const uint32_t h = neededPageCountForSide(extent.height, _subres.baseMipLevel+i);
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    uint32_t physPgAddr = phys_pg_addr_alctr_t::invalid_address;
+                    if (i>=levelsTakingAtLeastOnePageCount)
+                        physPgAddr = miptailPgAddr;
+                    else
+                    {
+                        const uint32_t szAndAlignment = 1u;
+                        core::address_allocator_traits<phys_pg_addr_alctr_t>::multi_alloc_addr(storage->tileAlctr, 1u, &physPgAddr, &szAndAlignment, &szAndAlignment, nullptr);
+                    }
+                    //assert(physPgAddr<SPhysPgOffset::invalid_addr);
+                    if (physPgAddr==phys_pg_addr_alctr_t::invalid_address)
+                    {
+                        free(offsetToTextureData(pgtOffset, _img, _wrapu, _wrapv), _img, _subres);
+                        return STextureData::invalid();
```

Comment:
let it fill the page table with the `physPgAddr` that is invalid, just skip the physical data copy